internal/handlers/common/aggregations/operators/type.go

```diff
@@ -56,28 +56,38 @@ func (t *typeOp) Process(doc *types.Document) (any, error) {
 		switch param := typeParam.(type) {
 		case *types.Document:
 			operator, err := NewOperator(param)
-			if errors.Is(err, ErrNoOperator) {
-				res = param
-				continue
-			}
-
 			if err != nil {
-				// TODO https://github.com/FerretDB/FerretDB/issues/2678
+				var opErr OperatorError
+				if !errors.As(err, &opErr) {
+					return nil, lazyerrors.Error(err)
+				}
+
+				if opErr.Code() == ErrNoOperator {
+					res = param
+					continue
+				}
+
 				return nil, err
 			}
 
 			if typeParam, err = operator.Process(doc); err != nil {
-				// TODO https://github.com/FerretDB/FerretDB/issues/2678
-				return nil, lazyerrors.Error(err)
+				var opErr OperatorError
+				if !errors.As(err, &opErr) {
+					return nil, lazyerrors.Error(err)
+				}
+
+				return nil, err
```

Comment:
This is not tested, please add a test when you remove todo comment.

Reply:
Is there a reason why this is not tested? It returns `OperatorError` so it should be tested right?

Reply:
I'm sure that after adding some recursive test cases, code coverage showed this line as covered. I believe that it's not covered right now, as I've moved error checking into the validation (it was necessary to work correctly with empty collections).

Reply:
Nice catch, it looks like all "Recursive" tests I've created weren't recursive :)